wasi/wasi_test.go

```diff
@@ -0,0 +1,199 @@
+			expectedArgs: [][]byte{
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+				[]byte("foobar\x00"),
+				[]byte("\x00"),
+				[]byte("baz\x00"),
+			},
+			expectedBufSize: 20,
+		},
+		{
+			name: "utf-8 string",
+			// "😨", "🤣", and "️🏃‍♀️" have 4, 4, and 13 bytes respectively
+			args: []string{"😨🤣🏃\u200d♀️", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("😨🤣🏃\u200d♀️\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 30,
+		},
+		{
+			name: "invalid utf-8 string",
+			args: []string{"\xff\xfe\xfd", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("\xff\xfe\xfd\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 12,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			argsOpt, err := Args(tc.args)
+			require.NoError(t, err)
+			wasiEnv := NewEnvironment(argsOpt)
+			hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+			// Serialize the expected result of args_size_get
+			argCountPtr := uint32(0)
+			expectedArgCount := make([]byte, 4)
+			binary.LittleEndian.PutUint32(expectedArgCount, uint32(len(tc.args)))
+			bufSizePtr := uint32(4)
+			expectedBufSize := make([]byte, 4)
+			binary.LittleEndian.PutUint32(expectedBufSize, tc.expectedBufSize)
+
+			// Compare them
+			errno := wasiEnv.args_sizes_get(hostFunctionCallContext, argCountPtr, bufSizePtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgCount, hostFunctionCallContext.Memory.Buffer[argCountPtr:argCountPtr+4])
+			require.Equal(t, expectedBufSize, hostFunctionCallContext.Memory.Buffer[bufSizePtr:bufSizePtr+4])
+
+			// Serialize the expected result of args_get
+			expectedArgs := make([]byte, 4*len(tc.args))
+			argsPtr := uint32(0)
+			expectedArgv := make([]byte, tc.expectedBufSize)
+			argvPtr := uint32(0x100)
+			argvWritten := uint32(0)
+			for i, arg := range tc.expectedArgs {
+				binary.LittleEndian.PutUint32(expectedArgs[argsPtr+uint32(i*4):], argvPtr+argvWritten)
+				copy(expectedArgv[argvWritten:], arg)
+				argvWritten += uint32(len(arg))
+			}
+
+			// Compare them
+			errno = wasiEnv.args_get(hostFunctionCallContext, argsPtr, argvPtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgs, hostFunctionCallContext.Memory.Buffer[argsPtr:argsPtr+uint32(len(expectedArgs))])
+			require.Equal(t, expectedArgv, hostFunctionCallContext.Memory.Buffer[argvPtr:argvPtr+uint32(len(expectedArgv))])
+		})
+	}
+}
+
+func TestArgsAPIReturnError(t *testing.T) {
+	dummyArgs := []string{"foo", "bar", "baz"}
+	argsOpt, err := Args(dummyArgs)
+	require.NoError(t, err)
+	wasiEnv := NewEnvironment(argsOpt)
+	hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+	outOfBounds := uint32(len(hostFunctionCallContext.Memory.Buffer))
+
+	argsArray, err := newWASIStringArray(dummyArgs)
+	require.NoError(t, err)
+
+	errno := wasiEnv.args_sizes_get(hostFunctionCallContext, outOfBounds, 0)
```

Comment:
we tend to use matrixes for tests that aren't inherently stateful as that allows you to name them. also it breaks it up visually a bit better than here. mind doing that and naming why the inputs are interesting? you can still re-use the setup above the matrix if helpful.

If these are stateful (between assertions), then you can comment why or do inline tests to name them after a previous. ex. `t.Run("second call fails", ...`

Reply:
Improved. Thanks!